vendors/st/boards/stm32l475_discovery/ports/common_io/iot_uart.c

```diff
@@ -0,0 +1,689 @@
+/**
+ * Amazon FreeRTOS V1.2.3
+ * Copyright (C) 2018 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
+
+/**
+ * @file iot_uart.c
+ * @brief file containing the implementation of UART APIs calling STM drivers.
+ */
+/* ST HAL API include */
+#include "stm32l475xx.h"
+#include "stm32l4xx_hal.h"
+#include "stm32l4xx_hal_uart.h"
+#include "stm32l475e_iot01.h"
+
+/* common-io UART include */
+#include "iot_uart.h"
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "semphr.h"
+
+/**
+ * @brief The number of USART ports supported by STM32l475_discovery board.
+ */
+#define IOT_USART_PORTS              ( ( uint32_t ) 5 )
+#define IOT_UART_BLOCKING_TIMEOUT    ( ( uint32_t ) 3000UL )
+#define IOT_UART_CLOSED              ( ( uint8_t ) 0 )
+#define IOT_UART_OPENED              ( ( uint8_t ) 1 )
+
+/**
+ * @brief STM UART Descriptor.
+ *
+ */
+typedef struct IotUARTDescriptor
+{
+    IotUARTCallback_t xUartCallback;
+    void * pvHalContext;
+    void * pvUserCallbackContext;
+    SemaphoreHandle_t xSemphr;
+    StaticSemaphore_t xSemphrBuffer;
+    uint8_t sOpened;
+} IotUARTDescriptor_t;
+
+/**
+ * @brief STM HAL Context abstracted in the IotUARTDescriptor_t.
+ *
+ * Contains all the parameters needed for working with STM driver.
+ */
+typedef struct
+{
+    UART_HandleTypeDef * pxHuart; /**< STM UART handle for all the configuration. */
+    IRQn_Type eIrqNum;            /**< Interrupt number for the UART. */
+} STM32_UART_HalContext_t;
+
+/**
+ * @brief Statically initialized map of STM UART Handle for all 5 ports.
+ *
+ */
+static UART_HandleTypeDef xUartHandleMap[] =
```

Comment:
This initialization could be done in approximately 80% fewer lines.

Reply:
Refactoring...